### PR TITLE
feat: suggest user defined memory variables in LG files

### LINKED ISSUE #6137: Wrong predictive text options shown for turn.activity
## Describe the bug

Autocomplete options are sometimes not helpful.

## Version

1.3.1-nightly.219396.61c944a

## To Reproduce

1. Open a Send a Response action
2. Add a new response to the LG file
3. Start typing a variable name (e.g. `turn.activity.text`)
4. Watch the autocomplete options; `activity` is correctly a field in `turn`, but then `text` is not found as a field on `activity`.

## Expected behavior

I would expect the options to be aware of the correct type of "activity" and offer suggestions based on it.

## Screenshots

image

### PR DESCRIPTION
**Pull request by cosmicshuai**
closes #6137

Now in LG all up view editor, user wrote some variables like `user.predefined.props`, the language server will suggest this variable when user try to enter it again.

Here is a short demo. (`user.group.name` is defined is dialog files)
userDefinedVa

Also in this PR, the calculation of memory variables has been refactored, this will reduce the calculation.

### CODE REVIEW

File: Composer/packages/tools/language-servers/language-generation/src/LGServer.ts [hunk 1]
```diff
@@ -817,6 +824,50 @@ export class LGServer {
...
+    if (projectId && fileId) {
+      const lgTextFiles = projectId ? this.getLgResources(projectId) : [];
+      const lgContents: string[] = [];
+      lgTextFiles.forEach((item) => {
+        if (item.id !== fileId) {
+          lgContents.push(item.content);
+        }
+      });
+
+      setTimeout(async () => {
```
Comment (hatpick) [inline Composer/packages/tools/language-servers/language-generation/src/LGServer.ts:838]: why the setTimeout?

File: Composer/packages/tools/language-servers/language-generation/src/lgParser.ts [hunk 2]
```diff
@@ -44,6 +44,19 @@ class LgParserWithoutWorker {
 
     return { suggestEntities: uniq(suggestEntities) };
   }
+
+  public async extractLGVariables(curCbangedFile: string | undefined, lgFiles: string[]) {
```
Comment (hatpick) [inline Composer/packages/tools/language-servers/language-generation/src/lgParser.ts:48]: does this need to be async?

File: Composer/packages/tools/language-servers/language-generation/src/LGServer.ts (hunk 1 above)
Comment (cosmicshuai) [reply Composer/packages/tools/language-servers/language-generation/src/LGServer.ts:841]: removed, it's not necessary.

File: Composer/packages/tools/language-servers/language-generation/src/lgParser.ts (hunk 2 above)
Comment (cosmicshuai) [reply Composer/packages/tools/language-servers/language-generation/src/lgParser.ts]: Removed async.

**Review by hatpick — APPROVED** [review]

**Comment by cosmicshuai** [comment]
Since this PR is related to add something in LG worker, it can wait until https://github.com/microsoft/BotFramework-Composer/pull/5853 is merged.

**Pull request merged**